src/rust-crypto/PerSessionKeyBackupDownloader.ts

```diff
@@ -0,0 +1,480 @@
+     * @param roomId - The room ID of the room where the error occurred.
+     * @param sessionId - The megolm session ID that is missing.
+     *
+     */
+    private async requestRoomKeyFromBackup(
+        version: string,
+        roomId: string,
+        sessionId: string,
+    ): Promise<KeyBackupSession> {
+        const path = encodeUri("/room_keys/keys/$roomId/$sessionId", {
+            $roomId: roomId,
+            $sessionId: sessionId,
+        });
+
+        return await this.http.authedRequest<KeyBackupSession>(Method.Get, path, { version }, undefined, {
+            prefix: ClientPrefix.V3,
+        });
+    }
+
+    private async downloadKeysLoop(): Promise<void> {
+        if (this.downloadLoopRunning) return;
+
+        // If we have a configuration problem, we don't want to try to download.
+        // If any configuration change is detected, we will retry and restart the loop.
+        if (this.hasConfigurationProblem) return;
+
+        this.downloadLoopRunning = true;
+
+        while (this.queuedRequests.length > 0) {
+            // we just peek the first one without removing it, so if a new request for same key comes in while we're
+            // processing this one, it won't queue another request.
+            const request = this.queuedRequests[0];
+            try {
+                const result = await this.queryKeyBackup(request.roomId, request.megolmSessionId);
+
+                if (this.stopped) {
+                    return;
+                }
+                // We got the encrypted key from backup, let's try to decrypt and import it.
+                try {
+                    await this.decryptAndImport(request, result);
+                } catch (e) {
+                    this.logger.error(
+                        `Error while decrypting and importing key backup for session ${request.megolmSessionId}`,
+                        e,
+                    );
+                }
+                // now remove the request from the queue as we've processed it.
+                this.queuedRequests.shift();
+            } catch (err) {
+                if (err instanceof KeyDownloadError) {
+                    switch (err.code) {
+                        case KeyDownloadErrorCode.MISSING_DECRYPTION_KEY:
+                            this.markAsNotFoundInBackup(request.megolmSessionId);
+                            // continue for next one
+                            this.queuedRequests.shift();
+                            break;
+                        case KeyDownloadErrorCode.NETWORK_ERROR:
+                            // We don't want to hammer if there is a problem, so wait a bit.
+                            await sleep(this.backoffDuration);
+                            break;
+                        case KeyDownloadErrorCode.STOPPED:
+                            // If the downloader was stopped, we don't want to retry.
+                            this.pauseLoop();
```

Comment:
it might be better to wrap the whole `while` loop in a `try { /*... */ } finally { this.pauseLoop(); }`, to make sure we don't forget to clear `backupLoopRunning`.